### PR TITLE
add tests on toolchain status with multiple members

### PR DESCRIPTION
**Pull request by xcoulon**
also, refactor the func to initialize the `cluster.GetMemberClustersFunc`
func by passing the name(s) of the expected member clusters.

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: pkg/controller/toolchainstatus/toolchainstatus_controller_test.go [hunk 1]
```diff
@@ -1101,19 +1162,25 @@ func newDeploymentWithConditions(deploymentName string, deploymentConditions ...
...
 }
 
-func newGetMemberClustersFuncEmpty(_ client.Client) cluster.GetMemberClustersFunc {
-	return func(conditions ...cluster.Condition) []*cluster.CachedToolchainCluster {
-		var clusters []*cluster.CachedToolchainCluster
-		return clusters
+func noMemberClusters() getMemberClustersFunc {
+	return func(cl client.Client) cluster.GetMemberClustersFunc {
+		return func(conditions ...cluster.Condition) []*cluster.CachedToolchainCluster {
+			return make([]*cluster.CachedToolchainCluster, 0)
```
Comment (xcoulon) [inline pkg/controller/toolchainstatus/toolchainstatus_controller_test.go:1182]: the return of the return of the return! 
Like in a low-budget movie sequel 😂

File: pkg/controller/toolchainstatus/toolchainstatus_controller_test.go [hunk 2]
```diff
@@ -1101,19 +1162,25 @@ func newDeploymentWithConditions(deploymentName string, deploymentConditions ...
...
+			return clusters
 		}
 	}
 }
 
-func newGetMemberClustersFuncEmpty(_ client.Client) cluster.GetMemberClustersFunc {
-	return func(conditions ...cluster.Condition) []*cluster.CachedToolchainCluster {
-		var clusters []*cluster.CachedToolchainCluster
-		return clusters
+func noMemberClusters() getMemberClustersFunc {
```
Comment (MatousJobanek) [inline pkg/controller/toolchainstatus/toolchainstatus_controller_test.go:1179]: do you really need this function - couldn't you just use `memberClusters()` as you did here: https://github.com/codeready-toolchain/host-operator/blob/761645b91eeeb0ea513c2879734fe85cea63e22e/pkg/controller/toolchainstatus/toolchainstatus_controller_test.go#L433

**Review by MatousJobanek — APPROVED** [review]
Looks good, but I guess that the function `noMemberClusters()` is not really needed

File: pkg/controller/toolchainstatus/toolchainstatus_controller_test.go (hunk 2 above)
Comment (xcoulon) [reply pkg/controller/toolchainstatus/toolchainstatus_controller_test.go:1179]: you're right. I thought that this `noMemberCluster()` func would be more expressive than something like `memberClusters()` (and I did not realize I used it elsewhere).
That being said, I'm fine with dropping `noMemberClusters()` and using `memberClusters()` instead. 
Let me know what you prefer.

Comment (MatousJobanek) [reply pkg/controller/toolchainstatus/toolchainstatus_controller_test.go:1179]: I prefer having just `memberClusters()` and dropping `noMemberCluster()`

Comment (xcoulon) [reply pkg/controller/toolchainstatus/toolchainstatus_controller_test.go]: ok, sounds good to me! 
done in 2c92725

File: pkg/controller/toolchainstatus/toolchainstatus_controller_test.go [hunk 3]
```diff
@@ -77,8 +77,8 @@ func prepareReconcile(t *testing.T, requestName string, httpTestClient *fakeHTTP
 	return r, reconcile.Request{NamespacedName: test.NamespacedName(test.HostOperatorNs, requestName)}, fakeClient
 }
 
-func prepareReconcileWithStatusConditions(t *testing.T, requestName string, conditions []toolchainv1alpha1.Condition, initObjs ...runtime.Object) (*ReconcileToolchainStatus, reconcile.Request, *test.FakeClient) {
-	reconciler, req, fakeClient := prepareReconcile(t, requestName, newResponseGood(), newGetMemberClustersFuncReady, initObjs...)
+func prepareReconcileWithStatusConditions(t *testing.T, requestName string, getMemberClustersFunc getMemberClustersFunc, conditions []toolchainv1alpha1.Condition, initObjs ...runtime.Object) (*ReconcileToolchainStatus, reconcile.Request, *test.FakeClient) {
```
Comment (rajivnathan) [inline pkg/controller/toolchainstatus/toolchainstatus_controller_test.go:80]: May be worth renaming either the variable or the type `getMemberClustersFunc` for readability

**Review by rajivnathan — APPROVED** [review]
Nice! Made a minor suggestion but it's only a suggestion.

File: pkg/controller/toolchainstatus/toolchainstatus_controller_test.go (hunk 3 above)
Comment (xcoulon) [reply pkg/controller/toolchainstatus/toolchainstatus_controller_test.go:80]: fair point, I'll see what I can do about it ;)

Comment (xcoulon) [reply pkg/controller/toolchainstatus/toolchainstatus_controller_test.go:80]: done in d37a2cd: code should be much simpler now:no more func with 3 levels of nested returns! 😅

**Review by MatousJobanek — APPROVED** [review]
Nice :+1:

**Review by rajivnathan — APPROVED** [review]
Awesome!

**Comment by xcoulon** [comment]
thanks for the feedback @MatousJobanek and @rajivnathan! 😊

**Pull request merged**